File: api/services/zerobounce.py
import os, logging
from typing import Any, Dict
import httpx
from .circuit_breaker import breakers

logger  = logging.getLogger("pullenspro")
API_KEY = os.getenv("ZEROBOUNCE_API_KEY", "")

# ZeroBounce status → our internal status
STATUS_MAP = {
    "valid":       "valid",
    "invalid":     "invalid",
    "catch-all":   "risky",
    "unknown":     "unknown",
    "spamtrap":    "invalid",
    "abuse":       "invalid",
    "do_not_mail": "invalid",
}

async def verify(email: str, client: httpx.AsyncClient) -> Dict[str, Any]:
    """
    7-tier email verification via ZeroBounce API.
    Covers: syntax, DNS/MX, domain existence, catch-all, role-based,
            disposable provider, SMTP verification + greylisting bypass.
    No raw SMTP ports required — ZeroBounce handles that server-side.
    """
    cb = breakers["zerobounce"]
    if not cb.allow_request():
        raise RuntimeError("ZeroBounce circuit breaker OPEN")
    try:
        r = await client.get(
            "https://api.zerobounce.net/v2/validate",
            params={"api_key": API_KEY, "email": email},
            timeout=20,
        )
        r.raise_for_status()
        data   = r.json()
        status = STATUS_MAP.get(data.get("status", "unknown"), "unknown")
        cb.record_success()
        return {
            "email":  email,
            "status": status,
            "score":  _score(data),
            "reason": data.get("sub_status") or data.get("status"),
            "checks": {
                "syntax":       data.get("status") != "invalid" or data.get("sub_status") != "mailbox_not_found",
                "mx_found":     data.get("mx_found") == "true",
                "smtp_provider": data.get("smtp_provider", ""),
                "catch_all":    data.get("catch_all") == "true",
                "role_based":   data.get("account", "").lower() in ROLE_ACCOUNTS,
                "disposable":   data.get("disposable") == "true",
                "free_email":   data.get("free_email") == "true",
            },
        }
    except Exception as exc:
        cb.record_failure()
        logger.error("ZeroBounce error: %s", exc)
        raise

def _score(data: Dict) -> int:
    s = data.get("status", "")
    if s == "valid":     return 95
    if s == "catch-all": return 55
    if s == "unknown":   return 30
    return 0

ROLE_ACCOUNTS = {
    "info", "contact", "admin", "support", "help", "sales",
    "hello", "noreply", "no-reply", "postmaster", "webmaster",
    "team", "hr", "careers", "jobs", "marketing", "billing",
}

File: api/services/zerobounce.js
import axios from "axios";
import { breakers } from "./circuitBreaker.js";

const API_KEY = process.env.ZEROBOUNCE_API_KEY || "";

const STATUS_MAP = {
  "valid":       "valid",
  "invalid":     "invalid",
  "catch-all":   "risky",
  "unknown":     "unknown",
  "spamtrap":    "invalid",
  "abuse":       "invalid",
  "do_not_mail": "invalid",
};

const ROLE_ACCOUNTS = new Set([
  "info","contact","admin","support","help","sales","hello",
  "noreply","no-reply","postmaster","webmaster","team","hr",
  "careers","jobs","marketing","billing","office","mail",
]);

export async function verify(email) {
  const cb = breakers.zerobounce;
  if (!cb.allowRequest()) throw new Error("ZeroBounce circuit breaker OPEN");
  try {
    const { data } = await axios.get("https://api.zerobounce.net/v2/validate", {
      params: { api_key: API_KEY, email },
      timeout: 20000,
    });
    const status = STATUS_MAP[data.status] || "unknown";
    const local  = email.split("@")[0]?.toLowerCase() || "";
    cb.recordSuccess();
    return {
      email,
      status,
      score: _score(data),
      reason: data.sub_status || data.status || null,
      checks: {
        mx_found:    data.mx_found === "true",
        smtp_provider: data.smtp_provider || "",
        catch_all:   data.catch_all === "true",
        role_based:  ROLE_ACCOUNTS.has(local),
        disposable:  data.disposable === "true",
        free_email:  data.free_email === "true",
      },
    };
  } catch (err) {
    cb.recordFailure();
    throw err;
  }
}

function _score(data) {
  const s = data.status || "";
  if (s === "valid")     return 95;
  if (s === "catch-all") return 55;
  if (s === "unknown")   return 30;
  return 0;
}
